### ISSUE TITLE
Syntax highlighting stops mid-paragraph

### ISSUE THREAD

**Issue by DOFfactory** [area/language-latex#132]
Hi, I just recently switched from Textmate. I like your plugin a lot, however, I am experiencing strange behaviour. 

In the middle of the paragraph it seems that the syntax highlighting stops. I initially thought that it is due to the quotation marks, but after deleting them nothing changes, there must be some other issue. Please have a look at the screenshot attached. 
![atom latex](https://user-images.githubusercontent.com/10846169/28459718-ec86ac5c-6e0e-11e7-8417-e8f879540617.png)
It simply mid-paragraph stops highlighting the `\ac{} ...` and `\cite{}` commands. 

On a more general note, I noticed that your package differentiates between 4 types of texts in the '.text.tex', 'text.bibtex', '.text.tex.latex', and 'text.log.latex'. Since I like syntax highlighting a lot, in Textmate I used to define my additional syntax highlighting for, e.g. everything within the `\footnote{...}`, or change the font size for `\section{...}` and `\subsection{...}`. Is it possible to do this with your package and introduce things like these?

Many thanks for your reply!

**Comment by Aerijo** [comment]
Hello, thanks for the question. The current package maintainers are unavailable currently, should be back mid August. In the mean time, maybe I can help.

When the highlighting stops, is it in a very long line? Try adding a lot of letters at the start to see if what is currently highlighted becomes plain. 

If this is the case you can put the line ```'limitLineLength': no``` in the grammar file (at the top; see https://github.com/area/language-latex/pull/131/files). I've submitted the pull request, just waiting for a collaborator to authorise it. Though changing your local copy may not be permanent if an update is released, hopefully the update has that line in it anyway.

If this is not the case, it may be that there are too many scoped tokens in your paragraph (check this by adding many ```\a```'s. This is a limitation of Atom, in that it will only look for a limited number of token matches per line, and my current workaround is to load the package https://atom.io/packages/grammar-token-limit. I set it to 150, but you can change it to suit your needs (of course, inserting a new line between sentences will fix it as well, and this does not affect the latex output).

For your second question, editing the grammar file is your best option. I'm not sure where to save edits (updates could erase them), but I'm sure you'll work something out. If you're unfamiliar with Atom grammar scoping, see the existing stuff and maybe this can guide you https://gist.github.com/DamnedScholar/622926bcd222eb1ddc483d12103fd315. Most important thing to note is that regex must have double ```\``` (eg. to find a match for ```\```, the regex is ```\\\\``` because normal regex is ```\\```). If you need more help, let me know.

**Comment by DOFfactory** [comment]
Thanks @Aerijo, this was helpful. I made both of the suggested changes, and I had to increase my token-limit to 300. It takes a while, but seems now everything is fine.

Regarding the grammar syntax highlighting, thanks for the link. Very helpful. Finally I start to understand how this all works, as I initially thought that all the additional changes should be made in the `style.less` file. Checking the syntax of this guide file, why the commands like `\section` etc. are not defined as foldables in this package (as suggested in the guide with the `foldingStartMarker` and `foldingStopMarker`?

**Comment by Aerijo** [comment]
I don't know, can't get it working myself. it only seems to recognise indentation folding, so try adding a space before subheading command, two spaces before actual text, etc.
```
\section{S1}
 This will be folded when section is folded, as will all lines starting one space or more in.
 \subsection{ss1}
  Text goes here. The spaces don't affect latex. Folds when subsection or section is folded.
 \subsection{ss2}
  You can fold this text separately.
```

**Comment by DOFfactory** [comment]
Yeah, I agree, pretty lame. This should be managed automatically by the syntax.

**Comment by Aerijo** [comment]
I'm not sure if it's an issue with the way I write it or not, but I can't get it to work with something even as simple as start = ```{``` and end = ```}```. This makes me think it's an issue with Atom itself, not the syntax, but I can do more testing and look at the major language packages, see how they handle it.

**Comment by Aerijo** [comment]
Not looking good: no folding rules for Python, C, Javascript or HTML. I can't offer a better solution for now other than indent using spaces.

**Comment by DOFfactory** [comment]
Well, watching this I think it should be relatively easy to code
[
![image](https://user-images.githubusercontent.com/10846169/28465734-1710d2ac-6e2a-11e7-9944-13034aee7dc4.png)
](url)

Taken the `(?:sub){0,2}section` for section  from the `latex.cson`, I start every (sub)section and (sub)paragraph to fold until the next (sub)section/(sub)paragraph, or `\end{document}` as a final.

**Comment by Aerijo** [comment]
My bad, turns out the curly braces I was testing with need to be escaped.

The following works for curly braces, customise it for (sub)?section stuff

```
foldingStartMarker: '\\{'
foldingStopMarker: '\\}'
```

By the way, I'm doing this testing in my own https://github.com/Aerijo/language-latex2e package. It's very much for personal use, as the scoping and colour scheme (for the odd scoping) is very crude and "angry fruit salad". Still, I like it because it taught me how to do all this and now I'm able to answer questions like this.

**Comment by Aerijo** [comment]
Still no good. If you have a lot of sections in one file, and still don't want to indent, maybe try using ```\input{path/to/file}``` to break it up. I just ```\input``` all the sections into a main document with mine, and tell the editor (using a magic comment) to compile that main file whenever I try and compile on a section file.

Also, see issues https://github.com/atom/atom/issues/11838 and https://github.com/atom/atom/issues/3442. Subscribing to them might let you know if a fix is released.

**Comment by DOFfactory** [comment]
Thanks for referencing these, now I understand this is something more serious than just a grammar issue. It seems that your tip to indent manually is the way to go for now. Looking forward for new development...

**Comment by Aerijo** [comment]
It's possible (and easy enough).

Add this to your stylesheet, changing the scoping and settings as desired. You can also add other customisations this way. Just be mindful that all scope names must be prepended by `syntax--` as shown.

```
atom-text-editor .syntax--text.syntax--tex.syntax--latex {
  .syntax--meta.syntax--function.syntax--section,
  .syntax--support.syntax--function.syntax--section {
    color: green;
    font-size: 16px;
    &.syntax--subsection {
        color: red;
    }
    &.syntax--subsubsection {
        color: violet;
    }
  }
}
```
The above settings will make all of the `\section` command green and 16px. It will then turn the *contents* red for a `subsection` (but otherwise same settings as parent), and violet for `subsubsection`. I don't (currently) know how to make the `\(sub)section` itself change colour too, as it's formatting is overridden by the scope `support.function.section.latex`, and this does not change when `sub`'s are added.

Having said this, Atom only sort of supports different fonts in the same file. It will make the font bigger, but **selecting text in that line and the one below behaves strangely**. I would reccomend not changing the font size, just the colour, but that's up to you.

**Comment by DOFfactory** [comment]
Many thanks @Aerijo for this. Much appreciated 🥇. I can second that the font resizing is not behaving properly. I even tried to set the `font-size: 150%;`, but it worked even worse than your code. I assume this is again Atom's behaviour, I tried to find a topic on this in the discussions, but no avail. Should we report this, or will they argue that this is a feature and not a bug? :-)

**Comment by Aerijo** [comment]
I don't think supporting multiple font sizes is a priority for Atom developers. I also don't need it myself, so I won't do anything about it.

However, you can open the issue yourself. I don't think they will call it a feature, as there is no reason for the bad selection to occur when the font is a different size; they already allow variable width fonts, so column alignment is hardly an issue. I have no idea how difficult it will be to fix.

If that's it though, I think it's time to close this issue. It's already wildly different from the original problem. Good luck if you choose to open an issue.

**Comment by DOFfactory** [comment]
Yeah, I searched for it again, and the only way to argue for it is for example using Markdown and when one defines a `#heading` and `##subheading`, to have it visibly distinguishable. Anyway, I will leave it for now.

I have another additional question. In your code, how do you know which scope to use? I have in mind `.syntax--meta.syntax--function.syntax--section,   .syntax--support.syntax--function.syntax--section`. I try to figure out whether `\footnote{}` has such a scope or not. I struggle to find the full list of available scopes. Is there any?

**Comment by Aerijo** [comment]
Windows, Linux, or Mac?

**Comment by DOFfactory** [comment]
Windows

**Comment by Aerijo** [comment]
Some combination of `shift, option, control, p` I think.

**Comment by DOFfactory** [comment]
Don't have an option key. I tried all the variations with shift, control, alt and P, nothing pops up.

**Comment by Aerijo** [comment]
I'm not sure, I use Mac. Put the cursor on the word you want to find it for, press the shortcut, and a popup will tell you the scopes of the current location.

Alternatively, the developer console can be used. Go into select element mode and hover over the text, it should have the scope in the `syntax--` form (it might have span or something at the start, ignore that).

**Comment by DOFfactory** [comment]
Ok, got it, thanks! For reference: Ctrl+Shift+P > type "scope" and it should show the first result `Log Cursor Scope`, hit enter

**Comment by Aerijo** [comment]
Does it have the windows shortcut next to it? That would be quicker and easier than opening command palette every time.

**Comment by DOFfactory** [comment]
Nope, trying to figure out how to set one :)

**Comment by Aerijo** [comment]
<img width="612" alt="screen shot 2017-07-23 at 8 54 00 pm" src="https://user-images.githubusercontent.com/22167388/28498779-1fe1dfa0-6fe9-11e7-8bc8-5e3ad149bec9.png">
This is what I see on Mac anyway. Sorry I can't help find / make the shortcut.

**Comment by DOFfactory** [comment]
Yeah, strange, Windows is showing this
![scope](https://user-images.githubusercontent.com/10846169/28498799-88a35476-6fa6-11e7-8bef-829c92c67953.png)

**Comment by Aerijo** [comment]
```
'atom-text-editor':
  'f2': 'editor:log-cursor-scope'
```
I amaze myself. Change `f2` to whatever key combination you want. Add this to your keymap file.

**Comment by DOFfactory** [comment]
Excellent, works now. Thanks!

**Comment by DOFfactory** [comment]
Hm, regarding syntax highlighting and indentation I came across of the [Beautify](https://github.com/Glavin001/atom-beautify) plugin. I did not try it yet, but there is a reference to [Latex Beautify](https://github.com/cmhughes/latexindent.pl) with some perl script.

**Comment by Aerijo** [comment]
This isn't really the place to have a conversation about things unrelated to the issue. Then again, I don't know anywhere else to do this.

Just got Beautify working. Seems OK. I know nothing about Perl though, so it was painful when it said I need to install yaml-tiny and some kind of unicode package.

**Comment by yudai-nkt** [comment]
Should be closed by #131
